File: selected_coding_interview/codes/java/lc_11_container_with_most_water/lc_11_container_with_most_water.java
/*
* File: lc_11_container_with_most_water.java
* Created Time: 2025-12-30
* Author: krahets
*/

package lc_11_container_with_most_water;

import include.*;
import java.util.*;

// ===== Solution Code =====
    class Solution {
        public int maxArea(int[] height) {
            int i = 0, j = height.length - 1, res = 0;
            while(i < j) {
                res = height[i] < height[j] ? 
                    Math.max(res, (j - i) * height[i++]): 
                    Math.max(res, (j - i) * height[j--]); 
            }
            return res;
        }
    }

public class lc_11_container_with_most_water {
    public static void main(String[] args) {
        // ======= Test Case =======
        // Test case 1
        int[] test_input = new int[]{1, 2, 3, 4, 5};

        // ====== Driver Code ======
        Solution slt = new Solution();
        var result = slt.maxArea(test_input);
        System.out.println(result);

    }
}

File: selected_coding_interview/codes/python/lc_11_container_with_most_water.py
"""
File: lc_11_container_with_most_water.py
Created Time: 2025-12-30
Author: krahets
"""

from include import *


# ===== Solution Code =====
class Solution:
    def maxArea(self, height: List[int]) -> int:
        i, j, res = 0, len(height) - 1, 0
        while i < j:
            if height[i] < height[j]:
                res = max(res, height[i] * (j - i))
                i += 1
            else:
                res = max(res, height[j] * (j - i))
                j -= 1
        return res


# ======= Test Case =======
# Test case 1
test_input = [1, 2, 3, 4, 5]

# ====== Driver Code ======
slt = Solution()
result = slt.maxArea(test_input)
print(result)
